src/core/leader_schedule.zig

```diff
@@ -9,205 +9,318 @@ const Pubkey = sig.core.Pubkey;
+            .leader_schedules_rw = RwMux(std.AutoArrayHashMap(Epoch, LeaderSchedule)).init(
+                std.AutoArrayHashMap(Epoch, LeaderSchedule).init(allocator),
+            ),
+        };
+    }
 
-    const Self = @This();
+    pub fn deinit(self: LeaderScheduleCache) void {
+        const leader_schedules, var leader_schedule_lg = self.leader_schedules_rw.writeWithLock();
+        defer leader_schedule_lg.unlock();
 
-    pub fn deinit(self: Self) void {
-        self.allocator.free(self.slot_leaders);
+        for (leader_schedules.values()) |v| v.deinit();
+        leader_schedules.deinit();
     }
 
-    pub fn getLeader(self: *const Self, slot: Slot) ?sig.core.Pubkey {
-        const index: usize = @intCast(slot - self.start_slot);
-        return if (index >= self.slot_leaders.len) null else self.slot_leaders[index];
+    pub fn getUniqueLeaders(self: *LeaderScheduleCache) !std.AutoArrayHashMap(Pubkey, void) {
+        const leader_schedules, var leader_schedules_lg = self.leader_schedules_rw.readWithLock();
+        defer leader_schedules_lg.unlock();
+
+        var unique_leaders = std.AutoArrayHashMap(Pubkey, void).init(self.allocator);
+
+        for (leader_schedules.values()) |leader_schedule| {
+            for (leader_schedule.slot_leaders) |leader| {
+                try unique_leaders.put(leader, void{});
+            }
+        }
+
+        return unique_leaders;
     }
 
-    pub fn provider(self: *Self) SlotLeaderProvider {
-        return SlotLeaderProvider.init(self, Self.getLeader);
+    pub fn getSlotLeaderProvider(self: *LeaderScheduleCache) SlotLeaderProvider {
+        return SlotLeaderProvider.init(self, LeaderScheduleCache.getSlotLeaderNoCompute);
     }
-};
 
-pub fn leaderScheduleFromBank(
-    allocator: Allocator,
-    bank_fields: *const sig.accounts_db.snapshots.BankFields,
-) !SingleEpochLeaderSchedule {
-    const epoch = bank_fields.epoch;
-    const epoch_stakes = bank_fields.epoch_stakes.getPtr(epoch) orelse return error.NoEpochStakes;
-    const slots_in_epoch = bank_fields.epoch_schedule.getSlotsInEpoch(epoch);
-    const staked_nodes = try epoch_stakes.stakes.vote_accounts.stakedNodes(allocator);
-
-    const slot_leaders = try leaderSchedule(allocator, staked_nodes, slots_in_epoch, epoch);
-
-    _, const slot_index = bank_fields.epoch_schedule.getEpochAndSlotIndex(bank_fields.slot);
-    const epoch_start_slot = bank_fields.slot - slot_index;
-    return SingleEpochLeaderSchedule{
-        .allocator = allocator,
-        .slot_leaders = slot_leaders,
-        .start_slot = epoch_start_slot,
-    };
-}
+    pub fn getSlotLeaderNoCompute(self: *LeaderScheduleCache, slot: Slot) ?Pubkey {
+        const epoch, const slot_index = self.epoch_schedule.getEpochAndSlotIndex(slot);
+        return self.getSlotLeader(epoch, slot_index);
+    }
+
+    pub fn getSlotLeaderMaybeCompute(self: *LeaderScheduleCache, slot: Slot, bank: *const BankFields) !sig.core.Pubkey {
```

Comment:
maybe better name is `getOrComputeSlotLeader`

Reply:
265ed12